### DISCUSSION TITLE
Unable to Start Java 17 Virtual Machine_M2 Mac

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#464] [Q&A] [answered]
I am unable to run test code on an Apple M2 Mac with the error message shown above .  I have the  py5 plugin installed on Thonny with the requisite JDK-17.  The HOME_PATH accurately reflects the location of JDK-17.  The installation has failed with the same error message on another M2 mac belonging to someone else, so I know that it is not unique to my system.  I have also installed 'Conda' thinking it would help, but do not know how to use it; it appears to be a cmd-line app run with Terminal.  Any advice would be appreciated.

Here's the entire error message:
```
py5 is unable to start a Java 17 Virtual Machine.
This library requires Java 17 to be installed and a properly set JAVA_HOME environment variable.
Here is some debug info about your installation that might help you identify the source of this problem.
{'JAVA_HOME environment variable': '/Users/xxxx/Library/Thonny/jdk-17', 'jvm version': (0, 0, 0), 'default jvm path': '/Users/xxxx/Library/Thonny/jdk-17/Contents/Home/lib/libjli.dylib'}
Traceback (most recent call last):
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/tools/run_sketch.py", line 58, in <module>
    main()
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/tools/run_sketch.py", line 49, in main
    imported.run_code(
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 148, in run_code
    _run_code(
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 304, in _run_code
    _run_sketch(sketch_path, classpath, exit_if_error)
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 226, in _run_sketch
    import py5
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5/__init__.py", line 75, in <module>
    raise RuntimeError("py5 is unable to start Java 17 Virtual Machine")
RuntimeError: py5 is unable to start Java 17 Virtual Machine

```

It appears to ultimately trace back to a function in the file 'py5_tools.jvm._start_jvm', **def _start_jvm()**.   If that helps any.

**Comment by hx2A** [comment]
It is true that others have had problems with py5 on the M2 machines but we've always been able to work through it to get it working. Can you tell me what version of py5 you have installed? If it is not the most current, can you upgrade? How did you install Java on this machine?

**Reply by vsquared** [reply]
It should be the latest version:

<img width="189" alt="py5" src="https://github.com/py5coding/py5generator/assets/2294268/adf66dbf-4f78-43fb-87c0-66ae4d4b2fcd">

**Reply by vsquared** [reply]
This machine is a clone of an older Intel chip machine using 'Time Machine', which as you know is just an external hard drive copy of the original.  I really can't tell you how Java was installed.  There are multiple jdk's on the system, upgraded over the years.  If I check _java --version_ from Terminal I am currently running openjdk 20.0.1.  ...But, py5 uses jdk-17 and it's in Thonny where it's supposed to be and HOME_PATH in the editor correctly points to it.  For whatever it's worth, the grandson has a brand new naive M2 system and when we tried to get py5 installed on it he got the same error message.

**Reply by hx2A** [reply]
> This machine is a clone of an older Intel chip machine using 'Time Machine', which as you know is just an external hard drive copy of the original.

Actually, I don't know that much about Macs but would think that a clone of an older Intel chip machine running on a M2 chip machine would have some problems here because it would be a different JVM...right? In any case, unraveling what is happening there will be complicated and is beyond the help that I can provide as an open source maintainer.

Getting it to work on the new machine is more achiveable. Can you evaluate that machine and make sure Java is installed correctly and that py5 is up to date? How was Java installed on that one?

Also note that I only have one of the older Intel Macs and not one of the new apple silicon machines. It would be great if someone else could help provide help for these questions because I cannot do so effectively.

**Reply by vsquared** [reply]
I appreciate your honesty.  No worries because I still have the Intel chip machine.  We'll keep working on it; there appears to be a solution somewhere, we just have to find it.  I'll keep looking.  For now we'll work on what we can do and not fret too much over what we can't do.  I appreciate the tremendous amount of time and effort that it takes for a project like this.  Thanks.

**Reply by hx2A** [reply]
The py5coding.org website has a [Special Notes for MacOS Users](http://py5coding.org/content/macos_users.html) and ideally I'd like to get all the folks with the Apple silicon computers to do knowledge sharing and augment that page to provide better advice for these setup issues.

**Comment by villares** [comment]
Hi @vsquared, I'm going to close this for now. I think the solution might be to (re)install a JDK built for Apple Silicon as reported at https://github.com/py5coding/py5generator/issues/452  

If you want to re-open this discussion, please create a new one naming the OS/Architecture on the subject, as it makes for more targeted discussion and attracts people with the same hardware.

**Reply by villares** [reply]
Hi @vsquared, I re-opened, as per your request. But please rename the subject adding it is about M2 Macs (I can't rename discussions, only issues)!

**Comment by vsquared** [comment, marked as answer]
I apologize for this long response but this is more complicated than I thought; bear with me.

One copy of jdk-17.jdk resides on the Macintosh HD at /Library/Java/JavaVirtualMachines.  If you have a brand new system there likely will be nothing in the JavaVirtualMachines folder.  If you used TimeMachine to 'clone' your old Intel mac there may be a jdk-17 folder in there, but it is for your old mac (_x86_64_) which will not work with the Thonny editor on a new M2 mac and you will get a '_py5 is unable to start Java 17 Virtual Machine_' error.  What you need for a new M2 mac system is a _jdk-17 for aarch64_.  To avoid confusion, it is probably best to delete the old jdk.  You can do that with the Terminal cmd-line using this code: **sudo rm -rf pathToFile**.  If you look into the JVM folder you should see that the old jdk-17 folder is now gone.  Next do a browser search for _download jdk 17 for mac_ and you should see an Oracle website where you can download jdk-17.  Look for a version for the newer macs with an M2 chip; it should be labelled **jdk-17_macos-aarch64_bin.dmg**.  When you have downloaded the .dmg file to your system click on it to setup the installer. The jdk will be automatically installed into Macintosh HD/Library/Java/JavaVirtualMachines; you can open the folder to check and see that the new jdk is now in place of the old one.  The next thing you have to do is edit the .zshrc file in Terminal to make certain that the JAVA_HOME path is set correctly.   On the cmd-line type vi .zshrc to open the file or create a new one if it doesn't already exist.   You should see the following at the top of the file: _JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home_; if it's not there go ahead and edit the file by typing 'i' which will allow you to insert it.  When you get all done typing hit the escape key ('esc') followed by '**:wq**' at the bottom of the file to save your changes.  You may navigate up and down in the **.zshrc** file with the arrow keys.  If you have 'conda' installed you may see their code in the file also; just leave it as is.  The only thing now left to do is set JAVA_HOME in the Thonny editor to match what you just entered into the .zshrc file.  Under the Tools menuItem on the toolbar select Options... Under the 'General' tab you should see a large white rectangle at the bottom for you to enter or copy/paste the same HOME_PATH that you entered in .zshrc.  'OK' the change and quit Thonny.  When you re-open Thonny you should no longer see the error message about being unable to start a java virtual machine.  I found the following YouTube video to be very helpful: https://www.youtube.com/watch?v=n15tOYj6d4E

**Addendum:**
There is also a Thonny folder at /Users/xxxx/Library/Thonny which will have jdk files in it.  This is a different Library folder than the one mentioned above.  I had to delete those jdks as well to finally get rid of the error message.  On a new system Apple may have hidden this Library menuItem.  Usually it is accessed with the Go menu in the Finder menubar.  If you don't see Library listed, then hold down either the control or option key while clicking on 'Go' and it should appear for you to click on.  There is a way to set it so it is always displayed by using the Terminal cmd-line but I don't recall what code is used; you should be able to find it with a browser search.  Unfortunately Thonny is programmed to add jdk-17 to this alternate location when trying to use _Imported mode for py5_, but in my limited experience on the M2 system it is the wrong jdk (needs further research); I still got the error message after Thonny told me everything was good to go.  At any rate I had to also get rid of all of those to finally eradicate the error message. **Oracle's jdk installer is putting the jdk in one place and Thonny is putting it in another (in addition it might not be the correct jdk for an M2 chip).**  Bottom line, we can't have it both ways and the _aarch64_ version of jdk-17 seems to work on my system.  I also got a request from Thonny to install a jdk when trying to use module mode after using imported mode; in my experience it's best to decline by hitting the Cancel button.  Thonny doesn't like it when there is no jdk where it expects one to be and I keep getting bombarded by requests to add the jdk.  I might have to move the jdk that Oracle installed (which is the correct one) to the Thonny folder and trick it into thinking everything is ok so it won't keep making the request to install one.  It seems like all of this is more difficult than it should be;  I thought I understood it, but there is more to it than I expected.

**Pertinent screenshots:**
<img width="589" alt="jdk_dwnld" src="https://github.com/py5coding/py5generator/assets/2294268/e0a9376d-07b1-4d76-977b-4db69cd38d4d">

<img width="113" alt="dmg" src="https://github.com/py5coding/py5generator/assets/2294268/f6afb0c7-9a3c-4bb8-a6e6-e5f2a751f101">

<img width="518" alt="Thonny_dlg" src="https://github.com/py5coding/py5generator/assets/2294268/3a384efa-0592-46a3-9d5e-104c0b559831">

**Addendum 2:**
To avoid having the Thonny dialog asking me to install jdk-17 I had to copy/paste the jdk (Oracle version) from the hard drive Library/Java/JavaVirtualMachines folder to /Users/xxxx/Library/Thonny and change the Thonny editor JAVA_HOME to point to this jdk.  Thonny apparently does not care about the .zshrc file JAVA_HOME path setting.

**Addendum 3:**
When I fired up my M2 this morning and tried to run a simple file in Thonny the error message was back, so I checked the HOME_PATH and found out that Thonny had added a line of code behind my back (see below).  That second line of code is incorrect;  when I deleted it the error message went away.  Now it looks like I'm going to have to change the jdk-17 filename by removing the .jdk extension just to make Thonny happy.

<img width="663" alt="PATH_HOME" src="https://github.com/py5coding/py5generator/assets/2294268/307c522a-9326-4406-b3d3-7a1aa10c30de">

**Reply by villares** [reply]
thank you so much for the detailed report!

I hope we can resume the Thonny py5 plugin development this year, and I think this contains valuable information to improve its behaviour.

Another idea we have been working on is finding someone with MacOS dev skills willing to re-package the Thonny IDE with py5 and the JDK pre-included, and that could mean a .dmg for Intel and another for Apple Silicon maybe... (just thinking).

**Reply by vsquared** [reply]
Having a Thonny editor with py5 and an appropriate jdk already installed in the correct place would be nice.  Please include jar files for JavaFX while you're at it.  I have no experience with .dmg creation or code signing for mac software; all the editors that I have written are for personal use, but I am willing to help as time permits.  Right now it's a struggle to use Thonny + py5 on Mac M2 and it's definitely different compared to the old Mac with an Intel chip.

**Comment by villares** [comment]
Moving this to https://github.com/tabreturn/thonny-py5mode/issues/55 as it seems to be a py5-mode plug-in issue!
